nitrite.go

```diff
@@ -204,289 +52,36 @@ func NewFetchingNitroCertProvider() (CertProvider, error) {
-	}
-
-	if x509.ECDSA != cert.PublicKeyAlgorithm {
-		return nil, ErrBadCertificatePublicKeyAlgorithm
-	}
-
-	if x509.ECDSAWithSHA384 != cert.SignatureAlgorithm {
-		return nil, ErrBadCertificateSigningAlgorithm
-	}
-
-	certificates = append(certificates, cert)
-
-	intermediates := x509.NewCertPool()
-
-	if !cert.IsCA {
-		for _, item := range doc.CABundle {
-			cert, err := x509.ParseCertificate(item)
-			if nil != err {
-				return nil, err
-			}
-
-			intermediates.AddCert(cert)
-			certificates = append(certificates, cert)
-		}
-	}
-
-	if verificationTime(doc).IsZero() {
-		return nil, fmt.Errorf("verification time is 0")
-	}
-
-	roots, err := certProvider.Roots()
-	if nil != err {
-		return nil, fmt.Errorf("getting root certificates: %w", err)
-	}
-
-	_, err = cert.Verify(
-		x509.VerifyOptions{
-			Intermediates: intermediates,
-			Roots:         roots,
-			CurrentTime:   verificationTime(doc),
-			KeyUsages: []x509.ExtKeyUsage{
-				x509.ExtKeyUsageAny,
-			},
-		},
-	)
-	if err != nil {
-		return nil, fmt.Errorf("verifying certificate: %w", err)
-	}
-
-	coseSig := coseSignature{
-		Context:     "Signature1",
-		Protected:   cose.Protected,
-		ExternalAAD: []byte{},
-		Payload:     cose.Payload,
-	}
-
-	sigStruct, err := cbor.Marshal(&coseSig)
-	if err != nil {
-		return nil, ErrMarshallingCoseSignature
-	}
+func WithTime(t time.Time) (VerificationTimeFunc, error) {
+	return func(_ internal.Document) time.Time {
+		return t
+	}, nil
+}
 
-	signatureOk := checkECDSASignature(
-		cert.PublicKey.(*ecdsa.PublicKey),
-		sigStruct,
-		cose.Signature,
+func NewDefaultVerifier() (*Verifier, error) {
+	return NewVerifier(
+		WithNitroCertProvider,
+		WithAttestationTime,
 	)
-
-	if !signatureOk {
-		err = ErrBadSignature
-	}
-
-	return &Result{
-		Document:     &doc,
-		Certificates: certificates,
-		Protected:    cose.Protected,
-		Unprotected:  cose.Unprotected,
-		Payload:      cose.Payload,
-		Signature:    cose.Signature,
-		SignatureOK:  signatureOk,
-		COSESign1:    sigStruct,
-	}, err
 }
 
-func checkECDSASignature(
-	publicKey *ecdsa.PublicKey,
-	sigStruct, signature []byte,
-) bool {
-	// https://datatracker.ietf.org/doc/html/rfc8152#section-8.1
-
-	var hashSigStruct []byte = nil
-
-	switch publicKey.Curve.Params().Name {
-	case "P-224":
-		h := sha256.Sum224(sigStruct)
-		hashSigStruct = h[:]
-
-	case "P-256":
-		h := sha256.Sum256(sigStruct)
-		hashSigStruct = h[:]
-
-	case "P-384":
-		h := sha512.Sum384(sigStruct)
-		hashSigStruct = h[:]
-
-	case "P-512":
-		h := sha512.Sum512(sigStruct)
-		hashSigStruct = h[:]
-
-	default:
-		panic(
-			fmt.Sprintf(
-				"unknown ECDSA curve name %v",
-				publicKey.Curve.Params().Name,
-			),
-		)
-	}
-
-	if len(signature) != 2*len(hashSigStruct) {
-		return false
+func (v *Verifier) Verify(attestation []byte) (*Result, error) {
+	result, err := internal.Verify(
+		attestation,
+		v.certProvider,
+		v.verificationTime,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("verifying attestation: %w", err)
```

Comment:
Just calling `Verify` for now, but we may be able to get rid of `internal.Verify` altogether after we refactor